src/voice/udp/handler.rs

```diff
@@ -0,0 +1,356 @@
+        let size = IpDiscoveryPacket::minimum_packet_size() + 64;
+
+        let mut buf: Vec<u8> = vec![0; size];
+
+        // Safety: expect is justified here, since this is an error which should never happen.
+        // If this errors, the code at fault is the buffer size calculation.
+        let mut ip_discovery_packet =
+            MutableIpDiscoveryPacket::new(&mut buf).expect("Mangled ip discovery packet creation buffer, something is very wrong. Please open an issue on the chorus github: https://github.com/polyphony-chat/chorus/issues/new");
+
+        ip_discovery_packet.populate(&ip_discovery);
+
+        let data = ip_discovery_packet.packet();
+
+        debug!("VUDP: Sending Ip Discovery {:?}", &data);
+
+        let send_res = udp_socket.send(data).await;
+        if let Err(e) = send_res {
+            return Err(VoiceUdpError::BrokenSocket {
+                error: format!("{:?}", e),
+            });
+        }
+
+        debug!("VUDP: Sent packet discovery request");
+
+        // Handle the ip discovery response
+        let received_size_or_err = udp_socket.recv(&mut buf).await;
+
+        if let Err(e) = received_size_or_err {
+            return Err(VoiceUdpError::BrokenSocket {
+                error: format!("{:?}", e),
+            });
+        }
+
+        let receieved_ip_discovery = IpDiscoveryPacket::new(&buf).expect("Could not make ipdiscovery packet from received data, something is very wrong. Please open an issue on the chorus github: https://github.com/polyphony-chat/chorus/issues/new");
+
+        debug!(
+            "VUDP: Received ip discovery!!! {:?}",
+            receieved_ip_discovery
+        );
+
+        let ip_discovery = IpDiscovery {
+            pkt_type: receieved_ip_discovery.get_pkt_type(),
+            length: receieved_ip_discovery.get_length(),
+            ssrc: receieved_ip_discovery.get_ssrc(),
+            address: receieved_ip_discovery.get_address(),
+            port: receieved_ip_discovery.get_port(),
+            payload: Vec::new(),
+        };
+
+        let mut data_reference_lock = data_reference.write().await;
+        data_reference_lock.ip_discovery = Some(ip_discovery);
+        drop(data_reference_lock);
+
+        let socket = Arc::new(udp_socket);
+
+        let events = VoiceUDPEvents::default();
+        let shared_events = Arc::new(Mutex::new(events));
+
+        let mut handler = UdpHandler {
+            events: shared_events.clone(),
+            data: data_reference.clone(),
+            socket: socket.clone(),
+        };
+
+        // Now we can continuously check for messages in a different task
+        tokio::spawn(async move {
+            handler.listen_task().await;
+        });
+
+        Ok(UdpHandle {
+            events: shared_events,
+            socket,
+            data: data_reference,
+        })
+    }
+
+    /// The main listen task;
+    ///
+    /// Receives udp messages and parses them.
```

Comment:
<!--
 Copyright (c) 2024 star
 
 This Source Code Form is subject to the terms of the Mozilla Public
 License, v. 2.0. If a copy of the MPL was not distributed with this
 file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->


```suggestion
    /// Receives UDP messages and parses them.
```